test/functional/Valhalla/src/org/openj9/test/lworld/ValueTypeUnsafeTests.java

```diff
@@ -190,4 +221,241 @@ static public void testuninitializedVTClassHasNullDefaultValue() {
+	}
+
+	@Test
+	static public void testGetValueOfZeroSizeVTObjectDoesNotCauseError() throws Throwable {
+		ZeroSizeValueTypeWrapper zsvtw = new ZeroSizeValueTypeWrapper();
+		long zsvtwOffset0 = myUnsafe.objectFieldOffset(ZeroSizeValueTypeWrapper.class.getDeclaredField("z"));
+		assertNotNull(myUnsafe.getValue(zsvtw, zsvtwOffset0, ZeroSizeValueType.class));
+	}
+
+	@Test
+	static public void testGetValuesOfObject() throws Throwable {
+		if (isFlatteningEnabled) {
+			ValueTypeInt x = myUnsafe.getValue(vtPoint, vtPointOffsetX, ValueTypeInt.class);
+			assertEquals(x.i, vtPoint.x.i);
+			ValueTypeInt y = myUnsafe.getValue(vtPoint, vtPointOffsetY, ValueTypeInt.class);
+			assertEquals(y.i, vtPoint.y.i);
+		}
+	}
+
+	@Test
+	static public void testGetValueOnVTWithLongFields() throws Throwable {
+		if (isFlatteningEnabled) {
+			ValueTypeLongPoint2D vtLongPoint = new ValueTypeLongPoint2D(123, 456);
+			ValueTypeLong x = myUnsafe.getValue(vtLongPoint, vtLongPointOffsetX, ValueTypeLong.class);
+			assertEquals(x.l, vtLongPoint.x.l);
+			ValueTypeLong y = myUnsafe.getValue(vtLongPoint, vtLongPointOffsetY, ValueTypeLong.class);
+			assertEquals(y.l, vtLongPoint.y.l);
+		}
+	}
+
+	@Test
+	static public void testGetValueOnNonVTObj() throws Throwable {
+		if (isFlatteningEnabled) {
+			IntWrapper i = new IntWrapper(7);
+			ValueTypeInt vti = myUnsafe.getValue(i, intWrapperOffsetI, ValueTypeInt.class);
+			assertEquals(vti.i, i.i);
+		}
+	}
+
+	@Test
+	static public void testNullObjPutValue() throws Throwable {
+		assertThrows(NullPointerException.class, () -> {
+			myUnsafe.putValue(null, vtPointOffsetX, ValueTypeInt.class, new ValueTypeInt(1));
+		});
+	}
+
+	@Test
+	static public void testNullClzPutValue() throws Throwable {
+		assertThrows(NullPointerException.class, () -> {
+			myUnsafe.putValue(vtPoint, vtPointOffsetX, null, new ValueTypeInt(1));
+		});
+	}
+
+	@Test
+	static public void testNullValuePutValue() throws Throwable {
+		assertThrows(NullPointerException.class, () -> {
+			myUnsafe.putValue(vtPoint, vtPointOffsetX, ValueTypeInt.class, null);
+		});
+	}
+
+	@Test
+	static public void testPutValueWithNonVTclzAndValue() throws Throwable {
+		int xBeforeTest = vtPoint.x.i;
+		myUnsafe.putValue(vtPoint, vtPointOffsetX, IntWrapper.class, new IntWrapper(10000));
+		assertEquals(vtPoint.x.i, xBeforeTest);
+	}
+
+	@Test
+	static public void testPutValueWithNonVTClzButVTValue() throws Throwable {
+		int xBeforeTest = vtPoint.x.i;
+		myUnsafe.putValue(vtPoint, vtPointOffsetX, IntWrapper.class, new ValueTypeInt(10000));
+		assertEquals(vtPoint.x.i, xBeforeTest);
+	}
+
+	@Test
+	static public void testPutValueWithVTClzButNonVTValue() throws Throwable {
+		if (isFlatteningEnabled) {
+			IntWrapper newVal = new IntWrapper(7392);
+			myUnsafe.putValue(vtPoint, vtPointOffsetX, ValueTypeInt.class, newVal);
+			int expectedValue = myUnsafe.getInt(newVal, myUnsafe.valueHeaderSize(ValueTypeInt.class));
```

Comment:
Is this behavior the same as OpenJDK ?

Reply:
yep, OpenJDK will let you do this as well

Reply:
OK, thanks.